target/mips/helper.c

```diff
@@ -1645,6 +1645,40 @@ bool mips_cpu_exec_interrupt(CPUState *cs, int interrupt_request)
 }
 
 #if !defined(CONFIG_USER_ONLY)
+bool r4k_lookup_tlb(CPUMIPSState *env, int *matching)
+{
+    bool mi = !!((env->CP0_Config5 >> CP0C5_MI) & 1);
+    r4k_tlb_t *tlb;
+    target_ulong mask;
+    target_ulong tag;
+    target_ulong VPN;
+    uint16_t ASID = env->CP0_EntryHi & env->CP0_EntryHi_ASID_mask;
+    uint32_t MMID = env->CP0_MemoryMapID;
+    uint32_t tlb_mmid;
+    int i;
+
+    MMID = mi ? MMID : (uint32_t) ASID;
+    for (i = 0; i < env->tlb->nb_tlb; i++) {
```

Comment:
`r4k_map_address` uses entries 0 to `env->tlb->tlb_in_use` not `nb_tlb`.  Do you mean to be comparing against entries that won't be matched?  You might, on the grounds that `tlb_in_use` might get increased later?  Is the rule for software that they have to erase entries above their intended `tlb_in_use` before decreasing it?

Reply:
I actually did not notice that `r4k_map_address` uses `env->tlb->tlb_in_use`, I set up the loop based on what the `r4k_helper_tlbp()` does. Isn't `tlb_in_use` always `>= nb_tlb`? It is my impression that the shadow TLB is above the fixed number of `nb_tlb` entries. Only entries from `tlbwr` end up in there, for which the patch adds a check to avoid duplicate insertion before the entry is evicted to shadow space. On a second thought, it might be the case that an entry that duplicates a shadow entry is added and then evicted by `tlbwr`, resulting in duplicate shadow entries.